### ISSUE TITLE
SPARC ELF binary not analyzed correctly

### ISSUE THREAD

**Issue by meowmeowxw** [radareorg/radare2#17850]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Arch Linux amd64
| File format of the file you reverse (mandatory)      | ELF
| Architecture/bits of the file (mandatory)            | SPARC 32 MSB
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.6.0-git 25177 @ linux-x86-64 git.4.4.0-842-g027992970 commit: 027992970402ef2ba84478bc5c8b433de8a08967 build: 2020-10-25__22:00:09

### Expected behavior

```
$ r2 ./sparc-1
[0x000100f0]>  aaa
[0x000100f0]> s main
[0x000104b4]> pdf # output taken from ghidra
0x000104b4 fc 3b a0 38     std        fp,[sp+local_res38]
0x000104b8 9c 03 bf a0     add        sp,-0x60,sp
0x000104bc bc 23 bf a0     sub        sp,-0x60,fp
0x000104c0 be 10 00 0f     mov        o7,i7
0x000104c4 7f ff ff 78     call       setup
0x000104c8 01 00 00 00     _nop
0x000104cc 7f ff ff 93     call       fn
0x000104d0 01 00 00 00     _nop
0x000104d4 82 10 20 00     mov        0x0,g1
0x000104d8 90 10 00 01     mov        g1,o0
0x000104dc 82 10 00 1e     mov        fp,g1
0x000104e0 9e 10 00 1f     mov        i7,o7
0x000104e4 fc 1f a0 38     ldd        [fp+local_res38],fp
0x000104e8 9c 10 00 01     mov        g1,sp
0x000104ec 81 c3 e0 08     retl
0x000104f0 01 00 00 00     _nop
```

### Actual behavior

```
$ r2 ./sparc-1
[0x000100f0]>  aaa
[0x000100f0]> s main
[0x000104b4]> pdf
p: Cannot find function at 0x000104b4
```

### Guess what's wrong

I tried to disassembly the main:

```
[0x000104b4]> pd
            ;-- main:
            0x000104b4      fc3ba038       invalid
            0x000104b5                    unaligned
            0x000104b6                    unaligned
            0x000104b7                    unaligned
            0x000104b8      9c03bfa0       add sp, -0x60, sp
            0x000104bc      bc23bfa0       sub sp, -0x60, fp
            0x000104c0      be10000f       mov o7, i7
            0x000104c4      7fffff78       call sym.setup
            0x000104c8      01000000       nop
            0x000104cc      7fffff93       call sym.fn
            0x000104d0      01000000       nop
            0x000104d4      82102000       mov 0, g1
            0x000104d8      90100001       mov g1, o0
            0x000104dc      8210001e       mov fp, g1
            0x000104e0      9e10001f       mov i7, o7
            0x000104e4      fc1fa038       invalid
            0x000104e5                    unaligned
            0x000104e6                    unaligned
            0x000104e7                    unaligned
            0x000104e8      9c100001       mov g1, sp
            0x000104ec      81c3e008       retl
            0x000104f0      01000000       nop
```

In ghidra this is correctly disassembled. I think that r2 expects to find the `save` and `restore` instructions in the prologue and epiloge of a function, or maybe it's just that the disassembler doesn't recognize the instructions. (I didn't look at the source code, it's just a guess).

This binary is from hack.lu ctf 2020, and it's hinted that is compiled with `-mflat`.
From the gcc's man:

```
With -mflat, the compiler does not generate save/restore instructions and uses a "flat" or single register window model.  This model is compatible with the regular register window model.  The local registers and the input registers (0--5) are still treated as "call-saved" registers and are
           saved on the stack as needed
```

### Additional Logs, screenshots, source-code,  configuration dump, ...

[sparc-1.zip](https://github.com/radareorg/radare2/files/5435971/sparc-1.zip)

**Comment by ret2libc** [comment]
I think it's because of capstone. Could you report this bug there?

```
$ ./cstool/cstool sparc fc3ba038
ERROR: invalid assembly code
```

This seems to happen both on master and on next branches.

**Comment by meowmeowxw** [comment]
sure!

**Comment by XVilka** [comment]
Would be also nice to have a test in radare2 after it's fixed.

**Comment by trufae** [comment]
fixed in 5.8.7

**Comment by trufae** [comment]
and ofc i added tests
